### PR TITLE
[CLIPPER-40] Clipper management client library

### PR DESCRIPTION
**Pull request by dcrankshaw**
Adds a Python library intended for use on a developer or data scientist's local machine (i.e. their laptop) to administer and manage a Clipper instance running in Docker. Includes functionality to start and stop an instance, add applications, models, and containers, and inspect various parts of the Clipper instance state. This version of the library interacts with the Clipper via SSH. [CLIPPER-80](https://clipper.atlassian.net/browse/CLIPPER-80) will add support for administering a local Clipper instance.

This functionality is needed for the demo described in [CLIPPER-73](https://clipper.atlassian.net/browse/CLIPPER-73).

### CODE REVIEW

File: management/clipper_manager.py [hunk 1]
```diff
@@ -0,0 +1,526 @@
...
+DOCKER_COMPOSE_DICT = {
+    'networks': {
+        'default': {
+            'external': {
+                'name': DOCKER_NW}}},
+    'services': {
+        'mgmt_frontend': {
+            'command': [
+                '--redis_ip=redis',
+                '--redis_port=6379'],
```
Comment (Corey-Zumar) [inline management/clipper_manager.py]: Let's define the relevant ports as constants at the top of the file.

File: management/clipper_manager.py [hunk 2]
```diff
@@ -0,0 +1,526 @@
...
+                fmt_result = dict([(splits[i], splits[i+1])
+                                   for i in range(0, len(splits), 2)])
+                pp = pprint.PrettyPrinter(indent=2)
+                pp.pprint(fmt_result)
+                return fmt_result
+            else:
+                warn("Application \"%s\" not found" % name)
+                return None
+
+    def inspect_selection_policy(self, app_name):
```
Comment (Corey-Zumar) [inline management/clipper_manager.py]: The parameter `uid` that is specified in the documentation is not included in the method prototype.

File: management/clipper_manager.py [hunk 3]
```diff
@@ -0,0 +1,526 @@
...
+        url = "http://%s:1338/admin/get_state" % self.host
+        req_json = json.dumps({
+            "app_name": app_name,
+            "uid": uid,
+        })
+        headers = {'Content-type': 'application/json'}
+        r = requests.post(url, headers=headers, data=req_json)
+        return r.text
+
+    def deploy_model(self, name, version, model, container, labels, num_containers=1):
```
Comment (Corey-Zumar) [inline management/clipper_manager.py]: change `container` to `container_name` for consistency with docs

File: management/clipper_manager.py [hunk 4]
```diff
@@ -0,0 +1,526 @@
...
+        container_name : str
+            The Docker container image to use to run this model container.
+        labels : list of str
+            A set of strings annotating the model
+        num_containers : int, optional
+            The number of replicas of the model to create. More replicas can be
+            created later as well. Defaults to 1.
+        """
+
+        # name = kwargs.pop('name')
```
Comment (Corey-Zumar) [inline management/clipper_manager.py]: Remove commented code

**Review by Corey-Zumar — APPROVED** [review]
Looks good overall. Because I don't have containers and applications to test, I've only verified that the following functions behave correctly:

- `start()`
- `list_apps()` (correctly indicates that no apps are present)
- `inspect_instance()`
- `cleanup()`
  - `stop_all()`

Just a couple of minor comments to be addressed, then LGTM.

File: management/clipper_manager.py (hunk 1 above)
Comment (dcrankshaw) [reply management/clipper_manager.py]: Done

File: management/clipper_manager.py (hunk 2 above)
Comment (dcrankshaw) [reply management/clipper_manager.py]: good catch. Forgot to delete kwargs.

File: management/clipper_manager.py (hunk 3 above)
Comment (dcrankshaw) [reply management/clipper_manager.py]: Done.

File: management/clipper_manager.py (hunk 4 above)
Comment (dcrankshaw) [reply management/clipper_manager.py]: Done

**Pull request merged**